factor/u2f/classes/factor.php

```diff
@@ -0,0 +1,251 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * U2F factor class (yubikey etc.).
+ *
+ * @package     factor_u2f
+ * @subpackage  tool_mfa
+ * @author      Jan Dageförde, Laura Troost
+ * @license     http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+namespace factor_u2f;
+
+defined('MOODLE_INTERNAL') || die();
+
+require_once($CFG->dirroot."/admin/tool/mfa/factor/u2f/thirdparty/vendor/autoload.php");
+
+use moodleform;
+use tool_mfa\local\factor\object_factor_base;
+use u2flib_server\Error;
+use u2flib_server\U2F;
+
+class factor extends object_factor_base {
+    /**
+     * User input for the generated code.
+     *
+     * {@inheritDoc}
+     */
+    public function login_form_definition($mform) {
+        global $PAGE, $CFG, $USER;
+
+        $renderer = $PAGE->get_renderer('core');
+        $pressbuttonhtml = $renderer->render_from_template('factor_u2f/press-button', [
+            'tokenillustration' => $renderer->image_url('token', 'factor_u2f'),
+        ]);
+        $mform->addElement('html', $pressbuttonhtml);
+        $mform->addElement('hidden', 'request', '', ["id" => 'id_request']);
+        $mform->setType('request', PARAM_RAW);
+        $mform->addElement('hidden', 'response_input', '', ['id'=> 'id_response_input']);
+        $mform->setType('response_input', PARAM_RAW);
+        $renderer = $PAGE->get_renderer('core');
+
+        $url = parse_url($CFG->wwwroot);
+        $u2f = new U2F($url['scheme'].'://'.$url['host']);
```

Comment:
its a small edge but multiples moodles on different paths will share the same $appId. If it works I'd just use the raw $CFG->wwwroot here, as far as I can tell $appId is just an arbitrary string and doesn't need to be a url?

Reply:
No, it has to be a URL and it may not include a path. https://developers.yubico.com/U2F/App_ID.html

However, the appID and the request object that are being created here are only valid for a specific request, so I see no issue with multiple moodles sharing the same appID.